src/cryptsetup/cryptsetup-token-systemd-tpm2/cryptsetup-token-systemd-tpm2.c

```diff
@@ -0,0 +1,283 @@
+
+        int r;
+        const char *json;
+        size_t blob_size, policy_hash_size, decrypted_key_size;
+        uint32_t pcr_mask;
+        systemd_tpm2_plugin_params params = {
+                .search_pcr_mask = UINT32_MAX
+        };
+        _cleanup_free_ void *blob = NULL, *policy_hash = NULL;
+        _cleanup_free_ char *base64_blob = NULL, *hex_policy_hash = NULL;
+        _cleanup_(erase_and_freep) void *decrypted_key = NULL;
+        _cleanup_(erase_and_freep) char *base64_encoded = NULL;
+
+        assert(token >= 0);
+
+        /* This must not fail at this moment (internal error) */
+        r = crypt_token_json_get(cd, token, &json);
+        assert(token == r);
+        assert(json);
+
+        if (usrptr)
+                params = *(systemd_tpm2_plugin_params *)usrptr;
+
+        r = parse_luks2_tpm2_data(json, params.search_pcr_mask, &pcr_mask, &base64_blob, &hex_policy_hash);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* should not happen since cryptsetup_token_validate have passed */
+        r = unbase64mem(base64_blob, SIZE_MAX, &blob, &blob_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* should not happen since cryptsetup_token_validate have passed */
+        r = unhexmem(hex_policy_hash, SIZE_MAX, &policy_hash, &policy_hash_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        r = acquire_luks2_key(
+                        pcr_mask,
+                        params.device,
+                        blob,
+                        blob_size,
+                        policy_hash,
+                        policy_hash_size,
+                        &decrypted_key,
+                        &decrypted_key_size);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* Before using this key as passphrase we base64 encode it, for compat with homed */
+        r = base64mem(decrypted_key, decrypted_key_size, &base64_encoded);
+        if (r < 0)
+                return log_debug_open_error(cd, r);
+
+        /* free'd automaticaly by libcryptsetup */
+        *password_len = strlen(base64_encoded);
+        *password = TAKE_PTR(base64_encoded);
+
+        return 0;
+}
+
+/*
+ * libcryptsetup callback for memory deallocation of 'password' parameter passed in
+ * any crypt_token_open_* plugin function
+ */
+_public_ void cryptsetup_token_buffer_free(void *buffer, size_t buffer_len) {
+        erase_and_free(buffer);
+}
+
+static int crypt_dump_buffer_to_hex_string(
+                const char *buf,
+                size_t buf_size,
+                char **ret_dump_str) {
+
+        int r;
+        unsigned int i;
+
+        for (i = 0; i < buf_size; i++) {
+                r = strextendf_with_separator(
+                        ret_dump_str,
+                        (i && !(i % 16)) ? CRYPT_DUMP_LINE_SEP : " ",
```

Comment:
This condition check is not obvious, please add a comment explaining why it's doing this

Reply:
sure, will add  it